examples/lock-app/linux/main.cpp

```diff
@@ -0,0 +1,65 @@
+/*
+ *
+ *    Copyright (c) 2020 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include <app/clusters/door-lock-server/door-lock-server.h>
+#include <app/util/af.h>
+
+#include "LockManager.h"
+
+#include "AppMain.h"
+
+using namespace chip;
+using namespace chip::app::Clusters;
+using namespace chip::app::Clusters::DoorLock;
+
+// App handles physical aspects of locking but not locking logic. That is it
+// should wait for door to be locked on lock command and return success) but
+// door lock server should check pin before even calling the lock-door
+// callback.
+bool emberAfPluginDoorLockOnDoorLockCommand(chip::EndpointId endpointId, const char * PINCode)
+{
+    return LockMgr().Lock(PINCode);
+}
+
+bool emberAfPluginDoorLockOnDoorUnlockCommand(chip::EndpointId endpointId, const char * PINCode)
+{
+    return LockMgr().Unlock(PINCode);
+}
+
+void MatterPostAttributeChangeCallback(const chip::app::ConcreteAttributePath & attributePath, uint8_t mask, uint8_t type, uint16_t size, uint8_t * value)
+{
+    // TODO: Watch for LockState, DoorState, Mode, etc changes and trigger appropriate action
+    if (attributePath.mClusterId == DoorLock::Id)
+    {
+        emberAfDoorLockClusterPrintln("Door Lock attribute changed");
+    }
+}
+
+void emberAfDoorLockClusterInitCallback(EndpointId endpoint)
```

Comment:
This should be in the cluster impl, I'd think...  And call into some sort of app-provided API for app-specific init if needed.

Reply:
Looking around at other apps all the *ClusterInitCallback() functions are defined in the app itself so I think this should be here. There is a MatterDoorLockPluginServerInitCallback() that is defined in door-lock-server.cpp that gets called on init.

Reply:
There are two different callbacks.  One is for stack startup init, one is for per-endpoint init.

That said, I just realized this is `emberAfDoorLockClusterInitCallback` which is indeed an app callback, unlike `emberAfDoorLockClusterServerInitCallback` which is the cluster-internal one.  We need better naming....